alvr/sockets/src/stream_socket/tcp.rs

```diff
@@ -0,0 +1,229 @@
+            .handle_try_again();
+
+        if res.is_ok() {
+            break;
+        }
+    }
+    let socket = res?.into();
+
+    crate::set_socket_buffers(&socket, send_buffer_bytes, recv_buffer_bytes).ok();
+    socket.set_read_timeout(Some(timeout)).to_con()?;
+
+    let socket = TcpStream::from(socket);
+
+    socket.set_nodelay(true).to_con()?;
+
+    Ok(socket)
+}
+
+pub struct MultiplexedTcpWriter {
+    inner: TcpStream,
+}
+
+impl MultiplexedSocketWriter for MultiplexedTcpWriter {
+    fn payload_offset(&self) -> usize {
+        PACKET_PREFIX_SIZE
+    }
+
+    // `buffer` contains the payload offset by `payload_offset()`
+    fn send(&mut self, stream_id: u16, packet_index: u32, buffer: &mut Vec<u8>) -> Result<()> {
+        let payload_size = buffer.len() - PACKET_PREFIX_SIZE;
+
+        buffer[0..2].copy_from_slice(&stream_id.to_le_bytes());
+        buffer[2..6].copy_from_slice(&packet_index.to_le_bytes());
+        buffer[6..10].copy_from_slice(&(payload_size as u32).to_le_bytes());
+
+        self.inner.write_all(buffer)?;
+
+        Ok(())
+    }
+}
+
+struct InProgressPacket {
+    stream_id: u16,
+    packet_index: u32,
+    buffer: Vec<u8>,
+    buffer_size: usize,
+    cursor: usize,
+}
+
+pub struct MultiplexedTcpReader {
+    inner: Socket,
+    in_progress_packet: Option<InProgressPacket>,
+    used_buffers_poll_timeout: Duration,
+}
+
+impl MultiplexedSocketReader for MultiplexedTcpReader {
+    fn payload_offset(&self) -> usize {
+        PACKET_PREFIX_SIZE
+    }
+
+    fn recv(&mut self, stream_queues: &HashMap<u16, StreamRecvQueues>) -> ConResult {
+        let in_progress_packet = if let Some(packet) = &mut self.in_progress_packet {
+            packet
+        } else {
+            let mut prefix_bytes = [0u8; PACKET_PREFIX_SIZE];
+
+            loop {
+                let count = self
+                    .inner
+                    .peek(unsafe {
+                        &mut *(&mut prefix_bytes as *mut [u8] as *mut [MaybeUninit<u8>])
+                    })
+                    .handle_try_again()?;
+                if count == PACKET_PREFIX_SIZE {
+                    break;
+                }
+            }
+
+            let stream_id = u16::from_le_bytes(prefix_bytes[0..2].try_into().unwrap());
+            let packet_index = u32::from_le_bytes(prefix_bytes[2..6].try_into().unwrap());
+            let payload_size = u32::from_le_bytes(prefix_bytes[6..10].try_into().unwrap()) as usize;
+
+            let mut buffer = match stream_queues.get(&stream_id) {
+                Some(queue) => queue
+                    .used_buffer_receiver
+                    .recv_timeout(self.used_buffers_poll_timeout)
+                    .handle_try_again()?,
+                None => {
+                    // This is a packet with an invalid stream id, but we must read it anyway. We
+                    // can't obtain a used buffer so we create a new one
+                    vec![]
+                }
```

Comment:
What's our design goal here? Pre-buffer up packets even if we're not yet actually having anything registered? Or try to waste as little on packets we don't need because we may have streams that get sent but not used? In the latter case we should make the buffer optional so we have the option to just dump the data (ofc figuring out how to not send the data in the first place would be optimal, but stuff like in-flight audio muting might make that quite hard)

Tho honestly since we're not actually buffering up anything except a single packet for the duration it's being actively received we should probably just implement some more efficient dumping (also probably not necessary, but why not, you like premature optimizations) (also likely out of scope)